src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -207,6 +207,77 @@ public static Array CreateInstance(Type elementType, params long[] lengths)
             return CreateInstance(elementType, intLengths);
         }
 
+        [RequiresDynamicCode("The code for an array of the specified type might not be available.")]
```

Comment:
The new APIs will not be annotated RequiresDynamicCode. These APIs being AOT safe was the explicit purpose for adding them (see "and requires suppressing warnings about the array type potentially not being available with Aot." in the issue).

The implementation currently generates a warning but that because of what jkotas said below: "That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be". We don't want the implementation to go back to array. It negates the whole point of the new API.

Reply:
Does it need `DynamicallyAccessedMembers.PublicConstructors` on the array type? That would trigger the problem I described.

Reply:
> The implementation currently generates a warning but that because of what jkotas said below: "That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be". We don't want the implementation to go back to array. It negates the whole point of the new API.

I think it is impossible in all cases 😥. I think in 99,9% of cases we can use `arrayType` to create an array but sometimes not. For example `arrayType` can be a type for a one-dimensional array and the user want to create a multi-dimensional array. So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

Reply:
> So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

I think the method should throw an exception if the dimensions don't match the array type.

It's the same as if we had a `ConstructorInfo` for a method that takes 1 parameter, but call it with two parameters. The BCL could in theory go looking for a different constructor on the type that matches the passed parameters. But it doesn't do that - it's an invalid use.